### PR TITLE
 JAVA-1755

### PR DESCRIPTION
**Pull request by jyemin**
Allow programmatic configurability of the StreamFactory for the async driver.

This needs some more tests, but first looking to see if a StreamFactoryFactory interface is the best approach.

    https://jira.mongodb.org/browse/JAVA-1755

### CODE REVIEW

**Comment by jyemin** [comment]
cc @craiggwilson

File: driver-core/src/main/com/mongodb/connection/netty/NettyStreamFactory.java [hunk 1]
```diff
@@ -47,7 +47,7 @@
      * @param eventLoopGroup the event loop group that all channels created by this factory will be a part of
      * @param allocator the allocator to use for ByteBuf instances
      */
-    public NettyStreamFactory(final SocketSettings settings, final SslSettings sslSettings, final NioEventLoopGroup eventLoopGroup,
+    public NettyStreamFactory(final SocketSettings settings, final SslSettings sslSettings, final EventLoopGroup eventLoopGroup,
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/connection/netty/NettyStreamFactory.java]: This is backwards breaking but better to do it now when no one is likely using it yet.

**Comment by jyemin** [comment]
ping @rozza @craiggwilson

**Comment by craiggwilson** [comment]
lgtm. My only concern is that you haven't provided a way for someone to wrap an existing StreamFactoryFactory to decorate it. Perhaps that isn't necessary in Java at all, in which case, ignore this.

**Comment by rozza** [comment]
lgtm

**Comment by jyemin** [comment]
What's the use case in the .NET driver for wrapping the StreamFactory?  When might someone want to do that?

**Comment by craiggwilson** [comment]
Socks5Stream, SshStream, etc... essentially protocol wrappers. If for some reason they wanted to GZip the stream, and had something on the other end to unzip before sending it to Mongo...

**Pull request closed without merging**